File: app-service-comments/src/main/java/com/github/arhor/aws/graphql/federation/comments/config/props/AppProps.java
package com.github.arhor.aws.graphql.federation.comments.config.props;

import jakarta.validation.Valid;
import jakarta.validation.constraints.Min;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import org.springframework.boot.context.properties.ConfigurationProperties;
import org.springframework.validation.annotation.Validated;

@Validated
@ConfigurationProperties(prefix = "app-props")
public record AppProps(
    @Valid
    @NotNull
    Events events,

    @NotNull
    @Min(0)
    Integer retryAttempts
) {
    public record Events(
        @Valid
        @NotNull
        Source source
    ) {
        public record Source(
            @NotBlank
            String syncCommentsOnPostCreatedEvent,

            @NotBlank
            String syncCommentsOnPostDeletedEvent,

            @NotBlank
            String syncCommentsOnUserCreatedEvent,

            @NotBlank
            String syncCommentsOnUserDeletedEvent
        ) {}
    }
}

File: app-service-scheduled-tasks/src/main/kotlin/com/github/arhor/aws/graphql/federation/scheduledTasks/config/props/AppProps.kt
package com.github.arhor.aws.graphql.federation.scheduledTasks.config.props

import jakarta.validation.Valid
import jakarta.validation.constraints.NotBlank
import jakarta.validation.constraints.NotNull
import org.springframework.boot.context.properties.ConfigurationProperties
import org.springframework.validation.annotation.Validated

@Validated
@ConfigurationProperties("app-props")
data class AppProps(
    @field:Valid
    @field:NotNull
    val events: Events?,
) {
    data class Events(
        @field:Valid
        @field:NotNull
        val target: Target?,

        @field:Valid
        @field:NotNull
        val source: Source?,
    ) {
        data class Target(
            @field:NotBlank
            val appEvents: String?,
        )

        data class Source(
            @field:NotBlank
            val createScheduledTaskEvents: String?,

            @field:NotBlank
            val deleteScheduledTaskEvents: String?,
        )
    }
}
